### PR TITLE
[JENKINS-68694] Winstone 6.1: Upgrade Jetty from 9.4.46.v20220331 to 10.0.11

### PR DESCRIPTION
**Pull request by basil**
See [JENKINS-68694](https://issues.jenkins.io/browse/JENKINS-68694). Jetty 10.0.x implements Servlet 4.0 (JakartaEE 8/`javax.servlet.*`). The minimum required Java version for Jetty 10 is now Java 11. For more details about the Jetty 10 release, see [this Jetty blog post](https://webtide.com/jetty-10-and-11-have-arrived/). Winstone 6.1 removes support for OpenSSL-style PEM-encoded RSA private keys in https://github.com/jenkinsci/winstone/pull/232.

### Testing done

- Generate a keystore as described in the [Jetty 10 Operations Guide](https://www.eclipse.org/jetty/documentation/jetty-10/operations-guide/index.html#og-keystore-create); successfully started Winstone with `--httpsKeyStore`, `--httpsKeyStorePassword`, and `--httpsPort`; visited the UI in my browser via TLS and clicked around successfully; build Freestyle and Pipeline jobs; verified Winstone 6.1 is displayed in the **About** page.
- [JENKINS-68933](https://issues.jenkins.io/browse/JENKINS-68933): Interactively verified `wsecho/` (in both text and binary modes, including ping visible after 30s with `websocat -vv`); Pipeline builds on a WebSocket agent; CLI in `-webSocket` mode.
- [JENKINS-68928](https://issues.jenkins.io/browse/JENKINS-68928): Ran `mvn -pl war jetty:run` 10 times in a row without any failures. Ensured that the `JENKINS_HOME` being used was the expected `$JENKINS_REPO/war/work` directory rather than `$HOME/.jenkins`.
- Ensured that the new `RealJenkinsRule`-based test added in #6780 failed without the addition of `Jetty10Provider` in this PR and passed with it.
- In https://github.com/jenkinsci/jenkins-test-harness/pull/453, https://github.com/jenkinsci/jenkins/pull/6802, and https://github.com/jenkinsci/bom/pull/1255, ran [the core test suite](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-6802/4/) and [BOM/PCT test suite](https://ci.jenkins.io/job/Tools/job/bom/job/PR-1255/5/) with a modified `JenkinsRule` based on Jetty 10. Without the addition of `Jetty10Provider` in this PR, these tests (including `jenkins.agents.WebSocketAgentsTest`) failed. With the addition of `Jetty10Provider` in this PR, all tests passed.
- Inspected `jenkins.war` before and after this change; verified that no unexpected JAR files had been added to or removed from the WAR archive.

### Proposed changelog entries

- Winstone 6.1: Upgrade Jetty from 9.4.46.v20220331 to 10.0.11 ([Winstone changelog](https://github.com/jenkinsci/winstone/releases/tag/winstone-6.1), [Jetty blog post](https://webtide.com/jetty-10-and-11-have-arrived/)).
- Remove support for OpenSSL-style PEM-encoded RSA private keys when running Jenkins with the embedded Jetty (Winstone) container and TLS (jenkinsci/winstone#232).

### Proposed upgrade guidelines

Support for OpenSSL-style PEM-encoded RSA private keys has been removed when running Jenkins with the embedded Jetty (Winstone) container and TLS. Specifically, the `--httpsPrivateKey` and `--httpsCertificate` flags have been removed in favor of the `--httpsKeyStore` flag. The removed flags have printed deprecation warnings since 2016 and were implemented with non-standard APIs that have since been removed from Java 17. The recommendation is to migrate to the `--httpsKeyStore` option, which takes a keystore as described in [the documentation](https://www.jenkins.io/doc/book/installing/initial-settings/#https-with-an-existing-certificate). As of [JEP 229](https://openjdk.org/jeps/229), [PKCS12](https://en.wikipedia.org/wiki/PKCS_12) is the recommended keystore type.

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since="TODO")` or `@Deprecated(since="TODO", forRemoval=true)` if applicable.
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6801"></a>

### CODE REVIEW

File: war/pom.xml
Comment (Vlatombe) [inline war/pom.xml:106]: It feels wrong to me to just add jetty 10 and not remove jetty 9.

Comment (basil) [reply war/pom.xml:106]: > It feels wrong to me

@vlatombe Whether it feels wrong to you or not (and I would suggest that you make requests based on explicit and objective reasoning rather than feelings), implicitly requesting that I remove Jetty 10 from `war/pom.xml` as you are doing here is not a request that can be accommodated (regardless of its underlying reasoning), because that would introduce test failures in any `JenkinsRule`-based test that uses WebSockets, such as `jenkins.agents.WebSocketAgentsTest` (because `JenkinsRule` still uses Jetty 9 and cannot be migrated to Java 11 and Jetty 10 until we drop support for older versions of Jenkins, even though that eventual migration has been tested and validated in https://github.com/jenkinsci/jenkins-test-harness/pull/453, https://github.com/jenkinsci/jenkins/pull/6802, and https://github.com/jenkinsci/bom/pull/1255). I suggest that you refrain from making implicit requests for change that would introduce test failures.

Comment (Vlatombe) [reply war/pom.xml:106]: I had missed the fact that `JenkinsRule` still relied on jetty 9. With that additional context, it looks perfectly fine.

File: test/src/test/java/hudson/security/csrf/CrumbExclusionTest.java
Comment (jglick) [inline test/src/test/java/hudson/security/csrf/CrumbExclusionTest.java:75]: or simply
```suggestion
                    assertThat("error message using " + path, x.getResponse().getContentAsString(), containsString("path parameter"));
```
?

File: test/src/test/java/jenkins/security/SuspiciousRequestFilterTest.java
Comment (jglick) [inline test/src/test/java/jenkins/security/SuspiciousRequestFilterTest.java:44]: similarly, could simplify to
```suggestion
        assertThat(response.getContentAsString(), containsString("path parameter"));
```
it seems

File: war/pom.xml
Comment (jglick) [inline war/pom.xml:617]: eclipse/jetty.project#7970 FTR

**Review by jglick — APPROVED** [review]

File: test/src/test/java/hudson/security/csrf/CrumbExclusionTest.java
Comment (basil) [reply test/src/test/java/hudson/security/csrf/CrumbExclusionTest.java:75]: I do not see a reason why we should simply do that.

Comment (jglick) [reply test/src/test/java/hudson/security/csrf/CrumbExclusionTest.java:75]: Well, because it is simpler and so easier to read, and a bit less likely to break with some future Jetty update.

**Review by jglick — APPROVED** [review]
Is this just awaiting another core reviewer to be ready to merge?

(Cheating a bit for me to review as a co-author.)

**Comment by basil** [comment]
> Is this just awaiting another core reviewer to be ready to merge?

Yes, it is.

File: test/src/test/java/hudson/security/csrf/CrumbExclusionTest.java
Comment (basil) [reply test/src/test/java/hudson/security/csrf/CrumbExclusionTest.java]: The readability argument seems subjective to me, but the fragility argument seems strong.

File: war/pom.xml
Comment (timja) [reply war/pom.xml:610]: Is there a workaround to this? i.e. without this we can't live reload resources from what I understand.

Comment (basil) [reply war/pom.xml:610]: I don't believe there is, but the pain should be short-lived as this has already been fixed upstream and should be present in the next release of Jetty.

**Review by timja — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
